crates/cli/src/subcommands/generate/csharp.rs

```diff
@@ -1,127 +1,533 @@
-    match ty {
-        // Options have a default value of null which is fine for us, and simple enums have their own default.
-        AlgebraicType::Sum(sum_type) if sum_type.is_option() || sum_type.is_simple_enum() => None,
-        // TODO: generate some proper default here (what would it be for tagged enums?).
-        AlgebraicType::Sum(_) => Some("null!"),
-        // For product types and arrays, we can use the default constructor.
-        AlgebraicType::Product(_) | AlgebraicType::Array(_) => Some("new()"),
-        // Strings must have explicit default value of "".
-        AlgebraicType::String => Some(r#""""#),
-        AlgebraicType::Ref(r) => default_init(ctx, &ctx.typespace[*r]),
-        _ => {
-            debug_assert!(ty.is_scalar());
-            None
-        }
+            writeln!(
+                output,
+                "public SubscriptionBuilder<EventContext> SubscriptionBuilder() => new(this);"
+            );
+        });
+
+        vec![("SpacetimeDBClient.g.cs".to_owned(), output.into_inner())]
+    }
+
+    fn clap_value() -> clap::builder::PossibleValue {
+        clap::builder::PossibleValue::new("csharp").aliases(["c#", "cs"])
     }
 }
 
-// can maybe do something fancy with this in the future
-fn csharp_typename(ctx: &GenCtx, typeref: AlgebraicTypeRef) -> &str {
-    ctx.names[typeref.idx()].as_deref().expect("tuples should have names")
+fn ty_fmt<'a>(module: &'a ModuleDef, ty: &'a AlgebraicTypeUse) -> impl fmt::Display + 'a {
+    fmt_fn(move |f| match ty {
+        AlgebraicTypeUse::Identity => f.write_str("SpacetimeDB.Identity"),
+        AlgebraicTypeUse::Address => f.write_str("SpacetimeDB.Address"),
+        AlgebraicTypeUse::ScheduleAt => f.write_str("SpacetimeDB.ScheduleAt"),
+        AlgebraicTypeUse::Unit => f.write_str("SpacetimeDB.Unit"),
+        AlgebraicTypeUse::Option(inner_ty) => write!(f, "{}?", ty_fmt(module, inner_ty)),
+        AlgebraicTypeUse::Array(elem_ty) => write!(f, "System.Collections.Generic.List<{}>", ty_fmt(module, elem_ty)),
+        AlgebraicTypeUse::String => f.write_str("string"),
+        AlgebraicTypeUse::Ref(r) => f.write_str(&type_ref_name(module, *r)),
+        AlgebraicTypeUse::Primitive(prim) => f.write_str(match prim {
+            PrimitiveType::Bool => "bool",
+            PrimitiveType::I8 => "sbyte",
+            PrimitiveType::U8 => "byte",
+            PrimitiveType::I16 => "short",
+            PrimitiveType::U16 => "ushort",
+            PrimitiveType::I32 => "int",
+            PrimitiveType::U32 => "uint",
+            PrimitiveType::I64 => "long",
+            PrimitiveType::U64 => "ulong",
+            PrimitiveType::I128 => "I128",
+            PrimitiveType::U128 => "U128",
+            PrimitiveType::I256 => "I256",
+            PrimitiveType::U256 => "U256",
+            PrimitiveType::F32 => "float",
+            PrimitiveType::F64 => "double",
+        }),
+        AlgebraicTypeUse::Never => unimplemented!(),
```

Comment:
I wonder if we could add a type to BSATN.Runtime that has no public constructors. That doesn't need to happen in this PR though. Not sure if there's any better way to translate the empty sum in C#.

Reply:
Yeah we can definitely do that - even abstract class would do - just didn't bother to in this PR. Happy to add one.

Reply:
Why would "never" type be ever in the module's type system though? If it can't be instantiated, it can't be (de)serialized either, right?

Reply:
@kazimuth did you mean to mark this as resolved?